### DISCUSSION TITLE
Support sharing text directly

### DISCUSSION THREAD

**Discussion by hyphenrf** [andyholmes/valent#682] [Feature Request]
I can receive text and links from kdeconnect android, but I can't find a dialogue that'll let me go the other way. Is that planned to be implemented?

**Comment by hyphenrf** [comment]
turns out I was on an old revision (nixos, 70ef1aa)

**Comment by hyphenrf** [comment]
I upgraded to the newest revision on master and confirmed that this request stands

**Comment by hyphenrf** [comment]
I found a new share dialogue, it was outside of the device view

**Reply by hyphenrf** [reply]
however, it only works with URLs and not plain text. as far as I can tell, there's no manual one-time "send clipboard" button which would cover for this feature somewhat

**Reply by andyholmes** [reply]
Sending text and syncing the clipboard are two different functions, but you're right that there's no way to share text directly.

AFAIK there's no UI function for this in KDE Connect either, and I'm unsure if there are plans for that. What is the use case for sharing text rather than a text file?

**Reply by hyphenrf** [reply]
not having to create an intermediate text file, really, and not having to receive it then open it with some application on the other end. I imagine the mechanism for sending urls is very similar to that of sending plaintext, and that the client does extra work to ensure the url is valid then specialize the protocol message with the correct type.
plaintext is usually for sharing a quick short note or even an ephemeral select-copied (middle-click pasted) passage of text. it's low-overhead and low-impact. I understand that clipboard sync is unrelated, but a "send-clipboard" button when clipboard-sync is turned off would cover this use-case somewhat. an even lower-overhead workflow would be to offer that functionality right from the gnome extension, so that one doesn't have to open a valent window. but that's going on too much of a tangent.
(having the ability to send plaintext also gives feature-symmetry with kdeconnect mobile, where you can highlight and share any text).

**Reply by andyholmes** [reply]
> I imagine the mechanism for sending urls is very similar to that of sending plaintext, and that the client does extra work to ensure the url is valid then specialize the protocol message with the correct type.

More or less, but the receiving mechanism is very poorly defined. Some implementations take the received text and dump it in a file, some show a dialog; Android just copies it to the clipboard.

> I understand that clipboard sync is unrelated, but a "send-clipboard" button when clipboard-sync is turned off would cover this use-case somewhat.

That's probably not a UI feature that will end up in Valent or the extension, as that didn't work out well at all in GSConnect. There are GAction's exported on D-Bus for that which can be used in third-party scripts and apps, though.

> having the ability to send plaintext also gives feature-symmetry with kdeconnect mobile, where you can highlight and share any text

This would be a useful feature, although our lack of a platform "Share" feature is really inconvenient. Currently the "Share" dialog is also basically a shim for that missing functionality, which currently looks like this:

![Screenshot of the Share Dialog, showing two options: files or URI](https://github.com/andyholmes/valent/assets/1265208/15844907-fa74-4a2e-8eb8-041153b4865c)

We could add a third row, which opens a text entry page similar to the "receive text" dialog:

![Screenshot of the Recieve Text dialog](https://github.com/andyholmes/valent/assets/1265208/00576807-7e3f-4f19-a152-d9e84df0b9cc)

There's a bit of work done on the `URI` entry (so that e.g. `www.gnome.org` can be assumed a valid URI and have `https://` prefixed), otherwise I'd just make that a magic "Enter URI or Text" field.

**Reply by hyphenrf** [reply]
Either way is good! "Enter URL" field doesn't resize though (aside: the "or paste text" part of the help text suggests pasting text does something, but it seems it doesn't?).
Maybe a user could type directly into it then have the option to open a larger dialogue when needed:
![image](https://github.com/andyholmes/valent/assets/26802762/d4f42431-7249-4890-8ffa-ea2773db88c3)
![image](https://github.com/andyholmes/valent/assets/26802762/bedcc461-88f5-48ce-a8c9-6f17e0cfaf45)
The go-next icon would serve as a way to open the share dialogue currently accessible only by hitting \<Enter\>, and in that dialogue, you could have a larger text box offering the ability to edit larger text and insert line-breaks. Although I fear there may be a bit of awkwardness with having two stages of input, maybe it's better to just drop the user right into the larger text box? or maybe reserve the smaller one to URLs and ask the user to go-next for plaintext? I don't really know and you definitely know better!

> as that didn't work out well at all in GSConnect

Could you give me context for this? I'm interested in understanding the limitations.
Also if there are protocol docs somewhere I wanna check them out :pray:
I tried digging for them on my own but found nothing official from KDE.

Also also, what ui mockup/preview tool do you use if any? I edited the above using inspector. As far as I can tell, the state of visually-editing ui files seems to be inadequate:
- `gtk4-builder-tool preview` fails on any ui file which has a template field
- `cambalache` requires a project-wide .cmb file, which many projects don't have. even when providing it that, it still struggles.
- `glade` doesn't support gtk4, `gnome-builder` uses `glade` under the hood (?)

**Reply by andyholmes** [reply]
> Could you give me context for this? I'm interested in understanding the limitations.

Trying to export an arbitrary amount of actions for device is problematic for a usable UI. An example in GSConnect is [#1640](https://github.com/GSConnect/gnome-shell-extension-gsconnect/issues/1640) , where the extension has to, but can't manage the amount of actions and items presented.

> Also if there are protocol docs somewhere I wanna check them out 🙏

The JSON RPC documentation is available at https://invent.kde.org/network/kdeconnect-meta/-/blob/master/protocol.md, although Valent maintains its own at https://valent.andyholmes.ca/documentation/protocol.html which are tested against.

> Also also, what ui mockup/preview tool do you use if any?

If I need a mock-up tool I just use [Workbench](https://github.com/workbenchdev/Workbench), although I usually just do it by hand in GNOME Builder when developing.

> Maybe a user could type directly into it then have the option to open a larger dialogue when needed:

I'd be more inclined to add a third row that opens a dedicated page with a large entry. I'm less worried about the small entry, than trying to differentiate a potential URI from arbitrary text (e.g. `www.gnome.org` vs `1.0.alpha`).

Something I'm less excited about it is relying on the behaviour of the remote device. Putting something into a file vs the clipboard could have quite different outcomes, maybe undesirable ones.

I haven't heard much feedback about this feature, so I don't know how/what it's being used for. @albertvaka, do you have any thoughts on the share/text behaviour? Should it be well-defined, or maybe just defined as "device may choose...".

**Reply by albertvaka** [reply]
There's some overlap between sharing text and clipboard sync. This is how we handle shared text it in the KDE desktop client:
- Copy the text to the clipboard and display a notification.
- The notification has an action to open the text in the default text editor.
- If the text contains only a link, the notification has a button to open the link.

I think is the best we can do, but for example in Android we don't do all this handling... so we are not consistent.

**Reply by hyphenrf** [reply]
it it because in android there's no notion of "default text editor"?

**Reply by albertvaka** [reply]
Yes, but we could still show a list of files to open with, or at least give the option to save as a file or something else. We just copy to the clipboard.

**Reply by andyholmes** [reply]
Hmm, a notification might be a better option for Valent, too. That's probably a decent expectation for behaviour, just that the receiving device will *notify* the user.
